### ISSUE TITLE
`explicit_auto_deref` lint suggests broken code

### ISSUE THREAD

**Issue by weiznich** [rust-lang/rust-clippy#9811]
### Summary

The `explicit_auto_deref` suggests broken code due to a false positive. For the provided diesel example the following changes are suggested:

```
warning: deref which would be done by auto-deref
   --> diesel/src/query_builder/ast_pass.rs:111:41
    |
111 |                 AstPassInternals::ToSql(*builder, options)
    |                                         ^^^^^^^^ help: try this: `builder`
    |
    = note: `#[warn(clippy::explicit_auto_deref)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#explicit_auto_deref

warning: deref which would be done by auto-deref
   --> diesel/src/query_builder/ast_pass.rs:117:28
    |
117 |                 collector: *collector,
    |                            ^^^^^^^^^^ help: try this: `collector`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#explicit_auto_deref

warning: deref which would be done by auto-deref
   --> diesel/src/query_builder/ast_pass.rs:118:34
    |
118 |                 metadata_lookup: *metadata_lookup,
    |                                  ^^^^^^^^^^^^^^^^ help: try this: `metadata_lookup`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#explicit_auto_deref
```

Applying these changes results in the following compiler error:

```
error[E0271]: type mismatch resolving `<DB as backend::Backend>::QueryBuilder == &mut <DB as backend::Backend>::QueryBuilder`
   --> diesel/src/query_builder/ast_pass.rs:111:17
    |
111 |                 AstPassInternals::ToSql(builder, options)
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected associated type, found mutable reference
    |
    = note: expected associated type `<DB as backend::Backend>::QueryBuilder`
             found mutable reference `&mut <DB as backend::Backend>::QueryBuilder`
    = help: consider constraining the associated type `<DB as backend::Backend>::QueryBuilder` to `&mut <DB as backend::Backend>::QueryBuilder` or calling a method that returns `<DB as backend::Backend>::QueryBuilder`
    = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html

error[E0271]: type mismatch resolving `<DB as backend::private::HasBindCollector<'_>>::BindCollector == &mut <DB as backend::private::HasBindCollector<'_>>::BindCollector`
   --> diesel/src/query_builder/ast_pass.rs:117:17
    |
117 |                 collector,
    |                 ^^^^^^^^^ expected associated type, found mutable reference
    |
    = note: expected associated type `<DB as backend::private::HasBindCollector<'_>>::BindCollector`
             found mutable reference `&mut <DB as backend::private::HasBindCollector<'_>>::BindCollector`
    = help: consider constraining the associated type `<DB as backend::private::HasBindCollector<'_>>::BindCollector` to `&mut <DB as backend::private::HasBindCollector<'_>>::BindCollector` or calling a method that returns `<DB as backend::private::HasBindCollector<'_>>::BindCollector`
    = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html

error[E0271]: type mismatch resolving `<DB as sql_types::TypeMetadata>::MetadataLookup == &mut <DB as sql_types::TypeMetadata>::MetadataLookup`
   --> diesel/src/query_builder/ast_pass.rs:118:17
    |
118 |                 metadata_lookup,
    |                 ^^^^^^^^^^^^^^^ expected associated type, found mutable reference
    |
    = note: expected associated type `<DB as sql_types::TypeMetadata>::MetadataLookup`
             found mutable reference `&mut <DB as sql_types::TypeMetadata>::MetadataLookup`
    = help: consider constraining the associated type `<DB as sql_types::TypeMetadata>::MetadataLookup` to `&mut <DB as sql_types::TypeMetadata>::MetadataLookup` or calling a method that returns `<DB as sql_types::TypeMetadata>::MetadataLookup`
    = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
```

### Lint Name

explicit_auto_deref

### Reproducer

Steps to reproduce: 

```
git clone https://github.com/weiznich/diesel
git checkout 523bd35be53299a5aee1df7e94680afb37d7cb06
cargo +1.65.0 clippy --tests --manifest-path diesel/Cargo.toml  --features "postgres"
# apply suggestions
cargo +1.65.0 clippy --tests --manifest-path diesel/Cargo.toml  --features "postgres"
# see error message
```

### Version

```text
rustc 1.65.0 (897e37553 2022-11-02)
binary: rustc
commit-hash: 897e37553bba8b42751c67658967889d11ecd120
commit-date: 2022-11-02
host: x86_64-unknown-linux-gnu
release: 1.65.0
LLVM version: 15.0.0
```

### Additional Labels

@rustbot label +I-suggestion-causes-error
